scipy/linalg/_cythonized_array_utils.pyx

```diff
@@ -66,34 +66,26 @@ def find_det_from_lu(lapack_t[:, ::1] a):
 
 
 # ====================== swap_c_and_f_layout : s, d, c, z ====================
-@cython.cdivision(True)
+@cython.nonecheck(False)
 @cython.wraparound(False)
 @cython.boundscheck(False)
 @cython.initializedcheck(False)
-cdef void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c, int n) noexcept nogil:
-    """Recursive matrix transposition for square arrays"""
-    cdef int i, j, ith_row, r2, c2
-    cdef lapack_t *bb=b
-    cdef lapack_t *aa=a
-    if r < 16:
-        for j in range(c):
-            ith_row = 0
-            for i in range(r):
-            # Basically b[i*n+j] = a[j*n+i] without index math
-                bb[ith_row] = aa[i]
-                ith_row += n
-            aa += n
-            bb += 1
-    else:
-        # If tall
-        if (r > c):
-            r2 = r//2
-            swap_c_and_f_layout(a, b, r2, c, n)
-            swap_c_and_f_layout(a + r2, b+(r2)*n, r-r2, c, n)
-        else:  # Nope
-            c2 = c//2
-            swap_c_and_f_layout(a, b, r, c2, n);
-            swap_c_and_f_layout(a+(c2)*n, b+c2, r, c-c2, n)
+cdef inline void swap_c_and_f_layout(lapack_t *a, lapack_t *b, int r, int c) noexcept nogil:
+    """
+    Swap+copy the memory layout of same sized buffers mainly
+    for Cython LAPACK interfaces.
+    """
+    cdef int row, col, ith_row
+    cdef lapack_t *bb = b
+    cdef lapack_t *aa = a
+
+    for col in range(c):
+        ith_row = 0
+        for row in range(r):
+            bb[row] = aa[ith_row]
+            ith_row += c
+        aa += 1
+        bb += r
```

Comment:
What's the reason for getting rid of the recursive descent here? Presumably this was added because it's faster to recurse into separate independent chunks (which don't hold the gil either, so can fully multithread) than going through one big loop?

I mean, I'm not against removing this (especially as performance seems to be better), but curious about this change.

Reply:
It was me writing that code under the impression that recursion helped but I did some experiments and it in fact slows down. One of those theory does not reflect to practice on modern hardware thing. 

Mostly correcting my too-smart self from the past.